### PR TITLE
Store per-difficulty metadata where it differs

### PR DESCRIPTION
**Pull request by Aergwyn**
I went ahead and created a PR to solve #1374. Just close if I was too eager here (regarding possible details I missed because I didn't understand the TODO in the code).

Now it displays the author per beatmap again.

### CODE REVIEW

**Comment by peppy** [comment]
Have you checked the mappings in the database for this? Do they reference the set-level metadata when there's no differences?

**Comment by Aergwyn** [comment]
Oh... no, will do later today.

**Comment by Aergwyn** [comment]
Oh. Forgot to mention. It did save for every beatmap even if duplicate (that's why it was set to null previously).
I implemented an Equals (for content equality, at least as far as possible) check and now it only stores metadata if it differs from the beatmapset metadata.

File: osu.Game/Beatmaps/BeatmapMetadata.cs [hunk 1]
```diff
@@ -66,5 +66,23 @@ public string AuthorString
             Source,
             Tags
         }.Where(s => !string.IsNullOrEmpty(s)).ToArray();
+
+        public override bool Equals(object other)
```
Comment (huoyaoyuan) [inline osu.Game/Beatmaps/BeatmapMetadata.cs]: Don't override `object.Equals`. Use `IEqutable<T>`.

File: osu.Game/Beatmaps/BeatmapMetadata.cs [hunk 2]
```diff
@@ -66,5 +66,23 @@ public string AuthorString
             Source,
             Tags
         }.Where(s => !string.IsNullOrEmpty(s)).ToArray();
+
+        public override bool Equals(object other)
+        {
+            var otherMetadata = other as BeatmapMetadata;
+            if (otherMetadata == null) return false;
+
+            return (onlineBeatmapSetID?.Equals(otherMetadata.onlineBeatmapSetID) ?? false)
```
Comment (huoyaoyuan) [inline osu.Game/Beatmaps/BeatmapMetadata.cs]: Use `operator ==` for primitive types.
_Defining custom `==` needs some more work, and seems bothering sometimes._

**Review by huoyaoyuan — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/BeatmapMetadata.cs (hunk 2 above)
Comment (Aergwyn) [reply osu.Game/Beatmaps/BeatmapMetadata.cs]: Oh right. I will do the same with PreviewTime then.

File: osu.Game/Beatmaps/BeatmapMetadata.cs (hunk 1 above)
Comment (Aergwyn) [reply osu.Game/Beatmaps/BeatmapMetadata.cs]: Good catch. Looking for references how this was done before I totally missed this detail. Will change.

Comment (huoyaoyuan) [reply osu.Game/Beatmaps/BeatmapMetadata.cs]: I did a lookup for all `object.Equals` references, and the only remained instance was `CultureInfo`.

File: osu.Game/Beatmaps/BeatmapMetadata.cs [hunk 3]
```diff
@@ -66,5 +67,23 @@ public string AuthorString
...
             Tags
         }.Where(s => !string.IsNullOrEmpty(s)).ToArray();
+
+        public bool Equals(BeatmapMetadata other)
+        {
+            if (other == null)
+                return false;
+
+            return onlineBeatmapSetID == other.onlineBeatmapSetID
+                && (Title?.Equals(other.Title) ?? false)
```
Comment (huoyaoyuan) [inline osu.Game/Beatmaps/BeatmapMetadata.cs]: I meant use `==` for all these lines 😦 
```csharp
=> other != null
&& onlineBeatmapSetID == other.onlineBeatmapSetID
.......
```

Comment (Aergwyn) [reply osu.Game/Beatmaps/BeatmapMetadata.cs]: Oh my god. Too much Java. Will fix.

**Comment by peppy** [comment]
We likely want to de-dupe in cases where multiple difficulties in the same set share metadata. Here's a few examples of what I mean:

sqlitestudio__3_1_1_

screen shot 2017-11-27 at 13 43 43

**Comment by Aergwyn** [comment]
Just preventing duplicates in that regard (I noticed that too, even nekodex on a fresh database sometimes creates 2 entries) or removing duplicates if existing?

**Comment by peppy** [comment]
Preventing is fine. We're still early in development so I'm not too concerned about migration/fixing databases when we can just re-import.

**Comment by Aergwyn** [comment]
This is the only location pre insert that I could get working to prevent duplicate entries. I'm not sure if I need to check the database too because if I saw correctly there has to be a BeatmapSet entry for a BeatmapMetadata entry and thus any import would return early anyways.
If it's wrong maybe it works as an inspiration how to do it?

**Comment by Aergwyn** [comment]
Everytime when I want to fall asleep I get weird thoughts:
Could it happen that metadata is equal on different beatmapsets? If yes: Does delete cascade?
In a theoretical scenario where both is answered with yes it could reuse metadata entries over multiple beatmapsets and may do some funny things once you delete one of them.

**Comment by peppy** [comment]
The de-duplication should *only be between beatmaps in the same set*.

**Comment by Aergwyn** [comment]
Yeah, what I thought. I just created the theory that I'd accidentally prune over different sets which I debunked myself (sneaky edits). So this proposed state would fulfill the requirements.

File: osu.Game/Beatmaps/BeatmapStore.cs [hunk 4]
```diff
@@ -32,6 +32,15 @@ public void Add(BeatmapSetInfo beatmapSet)
         {
             var context = GetContext();
 
+            foreach (var beatmap in beatmapSet.Beatmaps.Where(b => b.Metadata != null))
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapStore.cs]: i'm not sure we want to do this. we should be leaving the metadata field null where it inherits from its set, i think (so we can differentiate between set and beatmap metadata when edits occur).

Comment (Aergwyn) [reply osu.Game/Beatmaps/BeatmapStore.cs]: It is `null` though. The import makes sure of that.
This part just checks for duplicate metadata where metadata is present.
qq

Comment (peppy) [reply osu.Game/Beatmaps/BeatmapStore.cs]: So the `.Local` is checking yet-to-be-committed changes which is why this works? would definitely be worth adding a comment explaining what this is doing, as it's not immediately obvious.

File: osu.Game/Beatmaps/BeatmapStore.cs [hunk 5]
```diff
@@ -34,11 +34,12 @@ public void Add(BeatmapSetInfo beatmapSet)
 
             foreach (var beatmap in beatmapSet.Beatmaps.Where(b => b.Metadata != null))
             {
+                // check local context for metadata so we can reuse duplicates from the same set
                 var contextMetadata = context.Set<BeatmapMetadata>().Local.SingleOrDefault(e => e.Equals(beatmap.Metadata));
                 if (contextMetadata != null)
-                    beatmap.Metadata = contextMetadata;
+                    beatmap.Metadata = contextMetadata; // reuse existing
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapStore.cs]: unnecessary comment

File: osu.Game/Beatmaps/BeatmapStore.cs [hunk 6]
```diff
@@ -34,11 +34,12 @@ public void Add(BeatmapSetInfo beatmapSet)
 
             foreach (var beatmap in beatmapSet.Beatmaps.Where(b => b.Metadata != null))
             {
+                // check local context for metadata so we can reuse duplicates from the same set
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapStore.cs]: i was hoping for explanation on what `.Local` and `.Set` are doing here to achieve this..

**Review by peppy — APPROVED** [review]

**Comment by peppy** [comment]
looks good 👍

**Pull request merged**